### PR TITLE
[Event Hubs] Buffered Producer Client Infrastructure (enqueue)

### PR DESCRIPTION
**Pull request by jsquire**
# Summary

The focus of these changes is to continue fleshing out the buffered producer client, adding the ability to enqueue events to be published.

Pending work:

- Background management task
- Background publishing task
- Flush functionality
- Clear functionality
- Automatic partition assignment
- Partition Key hashing

# Related and References

- [EventHubBufferedProducerClient: Infrastructure and Functionality (#23480)](https://github.com/Azure/azure-sdk-for-net/issues/23480)

### CODE REVIEW

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs [hunk 1]
```diff
@@ -1101,7 +1265,7 @@ private async Task StartPublishingAsync(CancellationToken cancellationToken)
                 // If there is already a task running for the background management process,
                 // then no further initialization is needed.
 
-                if (IsPublishing)
+                if ((IsPublishing) && (!_producerManagementTask.IsCompleted))
```
Comment (christothes) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs:1268]: why the extra parans?

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClientOptions.cs [hunk 2]
```diff
@@ -16,57 +16,127 @@ namespace Azure.Messaging.EventHubs.Producer
...
+        /// <value>
+        ///   By default, idempotent retries are disabled.
+        /// </value>
+        ///
+        /// <remarks>
+        ///   It is important to note that enabling idempotent retries does not guarantee exactly-once semantics.  The existing
+        ///   Event Hubs at-least-once delivery contract still applies and event duplication is unlikely, but possible.
+        /// </remarks>
+        ///
+        public bool EnableIdempotentRetries { get; set; }
```
Comment (christothes) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClientOptions.cs:52]: Is it semantically correct to call this idempotent? Based on the description is attempting to de-duplicate retries. I think of "exactly-once" and idempotent distinct concepts. My understanding of an idempotent operation is that it is resilient to any number of duplicate invocations by guaranteeing that the effect of a single operation is equivalent to N of the same operation.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClientOptions.cs [hunk 3]
```diff
@@ -16,57 +16,127 @@ namespace Azure.Messaging.EventHubs.Producer
...
-        ///   <see cref="EventHubBufferedProducerClient.EnqueueEventsAsync(System.Collections.Generic.IEnumerable{EventData}, EnqueueEventOptions, System.Threading.CancellationToken)" />, which will automatically wait for room to be available.
+        ///   The total number of events that can be buffered for publishing at a given time for a given partition.  Once this capacity is reached, more events can enqueued by calling
+        ///   <see cref="EventHubBufferedProducerClient.EnqueueEventAsync(EventData, EnqueueEventOptions, System.Threading.CancellationToken)" /> or
+        ///   <see cref="EventHubBufferedProducerClient.EnqueueEventsAsync(IEnumerable{EventData}, EnqueueEventOptions, System.Threading.CancellationToken)" />, which will automatically wait for room to be available.
         /// </summary>
         ///
         /// <value>
-        ///   The default limit is 2,500 queued events.
+        ///   The default limit is 1,500 queued events for each partition.
```
Comment (christothes) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClientOptions.cs:90]: nit: the comma is culture specific.
```suggestion
        ///   The default limit is 1500 queued events for each partition.
```

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClientOptions.cs [hunk 4]
```diff
@@ -16,57 +16,127 @@ namespace Azure.Messaging.EventHubs.Producer
...
-        public int MaximumEventBufferLength { get; set; } = 2500;
+        /// <exception cref="ArgumentOutOfRangeException">Occurs when the requested count is not between 1 and 1,000,000 (inclusive).</exception>
+        ///
+        public int MaximumEventBufferLengthPerPartition
+        {
+            get => _maximumEventBufferLengthPerPartition;
+            set
+            {
+                // An upper bound of 1 million was chosen because it offers enough room for
+                // 2,147 partitions to be completely full before overflowing an integer value.
```
Comment (christothes) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClientOptions.cs:101]:
```suggestion
                // 2147 partitions to be completely full before overflowing an integer value.
```

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClientOptions.cs [hunk 5]
```diff
@@ -16,57 +16,127 @@ namespace Azure.Messaging.EventHubs.Producer
...
+                // The current cap for general Event Hubs use is 32 partitions.
+                //
+                // Important enterprise customers can request up to 2,000.
```
Comment (christothes) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClientOptions.cs:104]:
```suggestion
                // Important enterprise customers can request up to 2000.
```

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs (hunk 1 above)
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs:1268]: I prefer them for clarity, especially when additional decoration like negation is involved.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClientOptions.cs (hunk 2 above)
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClientOptions.cs:52]: Discussed offline.  Conceptually and semantically, I'm in agreement.  However, the nomenclature is matched to a service feature and needs to remain in alignment.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClientOptions.cs (hunk 3 above)
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClientOptions.cs:90]: Fair point.   I hadn't thought of that.

**Review by christothes — APPROVED** [review]
Force push be like:

**Pull request merged**